File: CartyObjC/CartyObjC/NativeViewController.m
//
//  NativeViewController.m
//  CartyObjC
//
//  Created by GZTD-03-01959 on 2026/1/5.
//

#import "NativeViewController.h"
#import <CartySDK/CartySDK.h>
#import "NativeTemplate.h"

@interface NativeViewController ()<CTNativeAdDelegate>


@property (nonatomic,weak)IBOutlet UIView *adView;
@property (nonatomic,weak)IBOutlet UILabel *infoLabel;
@property (nonatomic,strong)CTNativeAd *nativeAd;
@property (nonatomic,strong)NativeTemplate *nativeView;
@end

@implementation NativeViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.nativeAd = [[CTNativeAd alloc] init];
    self.nativeAd.placementid = @"281217152610";
    self.nativeAd.rootViewController = self;
    self.nativeAd.delegate = self;
}

-(IBAction)loadAd:(id)sender
{
    self.infoLabel.text = @"Start load Ad";
    [self.nativeAd loadAd];
}

-(IBAction)showAd:(id)sender
{
    if(!self.nativeAd.isReady)
    {
        self.infoLabel.text = @"Ad not Ready";
        return;
    }
    self.infoLabel.text = @"Start show Ad";
    if(self.nativeView != nil)
    {
        [self.nativeView removeFromSuperview];
    }
    self.nativeView = [[NSBundle mainBundle] loadNibNamed:@"NativeTemplate" owner:self options:nil].lastObject;
    [self.adView addSubview:self.nativeView];
    [self setFullLayOut:self.nativeView];
    [self.nativeView.mainImageView addSubview:self.nativeAd.mediaView];
    [self setFullLayOut:self.nativeAd.mediaView];
    [self.nativeView.adChoiceImageView addSubview:self.nativeAd.adChoiceView];
    [self setFullLayOut:self.nativeAd.adChoiceView];
    if(self.nativeAd.iconImageURL)
    {
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            NSData *data = [NSData dataWithContentsOfURL:[NSURL URLWithString:self.nativeAd.iconImageURL]];
            dispatch_async(dispatch_get_main_queue(), ^{
                self.nativeView.iconImageView.image = [UIImage imageWithData:data];
            });
        });
    }
    self.nativeView.titleLabel.text = self.nativeAd.title;
    self.nativeView.textLabel.text = self.nativeAd.desc;
    self.nativeView.ctaLabel.text = self.nativeAd.ctaText;
    [self.nativeAd registerContainer:self.nativeView withClickableViews:@[self.nativeView.titleLabel,self.nativeView.iconImageView,self.nativeView.mainImageView,self.nativeView.ctaLabel]];
}

- (void)setFullLayOut:(UIView *)view
{
    if(view.superview != nil)
    {
        view.translatesAutoresizingMaskIntoConstraints = false;
        NSLayoutConstraint *right = [NSLayoutConstraint constraintWithItem:view attribute:NSLayoutAttributeRight relatedBy:NSLayoutRelationEqual toItem:view.superview attribute:NSLayoutAttributeRight multiplier:1.0 constant:0];
        NSLayoutConstraint *left = [NSLayoutConstraint constraintWithItem:view attribute:NSLayoutAttributeLeft relatedBy:NSLayoutRelationEqual toItem:view.superview attribute:NSLayoutAttributeLeft multiplier:1.0 constant:0];
        NSLayoutConstraint *top = [NSLayoutConstraint constraintWithItem:view attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:view.superview attribute:NSLayoutAttributeTop multiplier:1.0 constant:0];
        NSLayoutConstraint *bottom = [NSLayoutConstraint constraintWithItem:view attribute:NSLayoutAttributeBottom relatedBy:NSLayoutRelationEqual toItem:view.superview attribute:NSLayoutAttributeBottom multiplier:1.0 constant:0];
        [view.superview addConstraints:@[right,left,top,bottom]];
    }
}

- (void)CTNativeAdDidLoad:(nonnull CTNativeAd *)ad { 
    self.infoLabel.text = @"Ad loaded";
}

- (void)CTNativeAdDidShow:(nonnull CTNativeAd *)ad { 
    self.infoLabel.text= @"Ad did show";
}

- (void)CTNativeAdLoadFail:(nonnull CTNativeAd *)ad withError:(nonnull NSError *)error { 
    self.infoLabel.text = [NSString stringWithFormat:@"Ad load fail %@",error];
}

- (void)CTNativeAdShowFail:(nonnull CTNativeAd *)ad withError:(nonnull NSError *)error { 
    self.infoLabel.text = [NSString stringWithFormat:@"Ad show fail %@",error];
}

- (void)CTNativeAdDidClick:(nonnull CTNativeAd *)ad {
    self.infoLabel.text = @"Ad click";
}
@end

File: CartySwift/CartySwift/NativeViewController.swift
//
//  NativeViewController.swift
//  CartySwift
//
//  Created by GZTD-03-01959 on 2026/1/5.
//

import UIKit
import CartySDK

class NativeViewController: UIViewController, CTNativeAdDelegate
{
    @IBOutlet var infoLabel:UILabel?
    @IBOutlet var adView:UIView?
    let nativeAd = CTNativeAd()
    var nativeView:NativeTemplate?
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        nativeAd.placementid = "281217152610"
        nativeAd.rootViewController = self
        nativeAd.delegate = self
    }
    
    @IBAction func loadAd()
    {
        infoLabel?.text = "Start load Ad"
        nativeAd.load()
    }
    
    @IBAction func showAd()
    {
        if(!nativeAd.isReady())
        {
            infoLabel?.text = "Ad not Ready"
            return
        }
        
        infoLabel?.text = "Start show Ad"
        if(nativeView != nil)
        {
            nativeView?.removeFromSuperview()
        }
        nativeView = (Bundle.main.loadNibNamed(
            "NativeTemplate",
            owner: nil,
            options: nil)?.first as! NativeTemplate)
        if(nativeView == nil)
        {
            return
        }
        adView?.addSubview(nativeView!)
        setFullLayOut(view: nativeView!)
        var array:[UIView] = NSMutableArray() as! [UIView];
        if(nativeView!.ctaLabel != nil)
        {
            nativeView!.ctaLabel!.text = nativeAd.ctaText
            array.append(nativeView!.ctaLabel!)
        }
        if(nativeView!.titleLabel != nil)
        {
            nativeView!.titleLabel!.text = nativeAd.title
            array.append(nativeView!.titleLabel!)
        }
        if(nativeView!.textLabel != nil)
        {
            nativeView!.textLabel!.text = nativeAd.desc;
            array.append(nativeView!.textLabel!)
        }
        if(nativeView!.iconImageView != nil
           && nativeAd.iconImageURL != nil)
        {
            fetchRemoteImage(urlStr: nativeAd.iconImageURL!)
            array.append(nativeView!.iconImageView!)
        }
        if(nativeAd.mediaView != nil)
        {
            nativeView!.mainImageView!.addSubview(nativeAd.mediaView!)
            setFullLayOut(view: nativeAd.mediaView!)
            array.append(nativeView!.mainImageView!)
        }
        if(nativeView!.adChoiceImageView != nil)
        {
            nativeView!.adChoiceImageView!.addSubview(nativeAd.adChoiceView!)
            setFullLayOut(view: nativeAd.adChoiceView!)
            nativeAd.adChoiceView!.frame = nativeView!.adChoiceImageView!.bounds;
            array.append(nativeView!.adChoiceImageView!)
        }
        nativeAd.registerContainer(nativeView!, withClickableViews: array)
    }
    
    func setFullLayOut(view:UIView)
    {
        if(view.superview != nil)
        {
            view.translatesAutoresizingMaskIntoConstraints = false;
            let right = NSLayoutConstraint(item: view, attribute: .right, relatedBy:.equal, toItem:view.superview, attribute:.right, multiplier:1.0, constant: 0 )
            let left = NSLayoutConstraint(item: view, attribute: .left, relatedBy:.equal, toItem:view.superview, attribute:.left, multiplier:1.0, constant: 0 )
            let top = NSLayoutConstraint(item: view, attribute: .top, relatedBy:.equal, toItem:view.superview, attribute:.top, multiplier:1.0, constant: 0 )
            let bottom = NSLayoutConstraint(item: view, attribute: .bottom, relatedBy:.equal, toItem:view.superview, attribute:.bottom, multiplier:1.0, constant: 0 )
            view.superview?.addConstraints([right,left,top,bottom])
        }
    }
    
    func fetchRemoteImage(urlStr:String)
    {
        guard let url = URL(string: urlStr) else { return }
        URLSession.shared.dataTask(with: url){ (data, response, error) in
            if let image = UIImage(data: data!){
                DispatchQueue.main.async { [self] in
                    nativeView!.iconImageView?.image = image
                }
            }
            else{
                print(error ?? "")
            }
        }.resume()
    }
    
    func ctNativeAdDidLoad(_ ad: CTNativeAd) {
        infoLabel?.text = "Ad loaded"
    }
    
    func ctNativeAdLoadFail(_ ad: CTNativeAd, withError error: any Error) {
        infoLabel?.text = "Ad load fail \(error)"
    }
    
    func ctNativeAdDidShow(_ ad: CTNativeAd) {
        infoLabel?.text = "Ad did show"
    }
    
    func ctNativeAdShowFail(_ ad: CTNativeAd, withError error: any Error) {
        infoLabel?.text = "Ad show fail \(error)"
    }
    
    func ctNativeAdDidClick(_ ad: CTNativeAd) {
        infoLabel?.text = "Ad click"
    }
}
